### PR TITLE
bpo-37382: check for debug info before testing GDB

### PR DESCRIPTION
**Pull request by jdemeyer**
https://bugs.python.org/issue37382

### CODE REVIEW

**Comment by jdemeyer** [comment]
CC @encukou

**Comment by jdemeyer** [comment]
Bugfix, so should be backported to all earlier versions.

**Review by vstinner — CHANGES_REQUESTED** [review]
The tests should be skipped if Python is optimized: see the issue.

**Comment by jdemeyer** [comment]
So how would you suggest to fix bpo-37382?

**Comment by vstinner** [comment]
https://bugs.python.org/issue37382 is not a bug.

**Comment by jdemeyer** [comment]
At least part of it is a bug: `test_gdb` fails when compiling CPython with `-O0 -g0`.

But the real issue for me is that `test_gdb` is not run by default. This PR addresses that.

File: Lib/test/support/__init__.py [hunk 1]
```diff
@@ -1643,6 +1643,16 @@ def disable_gc():
             gc.enable()
 
 
+def python_has_debug_info():
+    """Find if Python was built with debug information."""
+    cflags = sysconfig.get_config_var('PY_CFLAGS') or ''
+    final_opt = ""
+    for opt in cflags.split():
+        if opt.startswith('-g'):
+            final_opt = opt
```
Comment (vstinner) [inline Lib/test/support/__init__.py:1652]: Why not just returning True here? (and "return False" outside the loop)

File: Lib/test/support/__init__.py [hunk 2]
```diff
@@ -1643,6 +1643,16 @@ def disable_gc():
...
+    return final_opt == '-g'
```
Comment (vstinner) [inline Lib/test/support/__init__.py:1653]: Why do you reject "-ggdb"? I heard that it's even better than "-g".

File: Doc/library/test.rst [hunk 3]
```diff
@@ -414,6 +414,11 @@ The :mod:`test.support` module defines the following functions:
    Return ``True`` if Python was not built with ``-O0`` or ``-Og``.
 
 
+.. function:: python_has_debug_info()
+
+   Return ``True`` if Python was built with ``-g``.
```
Comment (vstinner) [inline Doc/library/test.rst:419]: Add ".. versionadded:: 3.9".

**Review by vstinner — COMMENTED** [review]
Please don't backport this change to Python 3.8 :-) test_gdb is giving me too many headaches :-) For example, it fails if Python is built with -Os.

**Comment by encukou** [comment]
Sorry for missing the ping back in 2019.
This PR has been superseded by [#110026](https://github.com/python/cpython/pull/110026/files#diff-65a7e7fedffb3771932da22df1aa9c74f8ec3781ed33775db45ac011be17e8a0) (and others). 
The current settings probably skip more than they need to, so the issue might still be relevant -- but a PR for that should now skip fewer tests rather than more.

**Pull request closed without merging**